service/history/handler.go

```diff
@@ -703,6 +704,42 @@ func (h *handlerImpl) StartWorkflowExecution(
 	return response, nil
 }
 
+func (h *handlerImpl) DescribeShardDistribution(
+	ctx context.Context,
+	request *types.DescribeShardDistributionRequest,
+) (resp *types.DescribeShardDistributionResponse, retError error) {
+	numShards := h.config.NumberOfShards
+	resp = &types.DescribeShardDistributionResponse{
+		NumberOfShards: 0,
+		Shards:         make(map[int32]string),
+	}
+	if len(*request.Role) == 0 {
+		*request.Role = "cadence-history"
+	}
+
+	offset := request.PageID * request.PageSize
+	// count all the shards but put only the ones falling into the requested page
+	addShardToResponse := func(shardID int, identity string) {
+		if resp.NumberOfShards >= offset && len(resp.Shards) < int(request.PageSize) {
+			resp.Shards[int32(shardID)] = identity
+		}
+		resp.NumberOfShards++
+	}
+
+	for shardID := 0; shardID < numShards; shardID++ {
+		info, err := h.GetHistoryServiceResolver().Lookup(string(rune(shardID)))
```

Comment:
I think both `HistoryServiceResolver` and `numberOfHistoryShards` (number of shards is in the static yaml config) are available in the frontend. So it looks like we can handle the request directly in frontend instead of forwarding it to history service.

Reply:
Let me try and see if this works

Reply:
The config on admin side has `NumHistoryShards`; is this just history shards. If so, do we have access to total number of shards of all types ?

Reply:
Talked offline this is now addressed